DirectProgramming/DPC++FPGA/Tutorials/Tools/dynamic_profiler/src/dynamic_profiler.cpp

```diff
@@ -0,0 +1,239 @@
+class ConsumerAfterKernel;
+
+// kSize = # of floats to process on each kernel execution.
+#if defined(FPGA_EMULATOR)
+constexpr int kSize = 4096;
+#else
+constexpr int kSize = 262144;
+#endif
+
+// Number of iterations performed in the consumer kernels
+// This controls the amount of work done by the Consumer.
+constexpr int kComplexity = 2000;
+
+// Perform two stages of processing on the input data.
+// The output of ConsumerWork1 needs to go to the input
+// of ConsumerWork2, so they cannot be done concurrently.
+// These functions are currently doing pointless work, but 
+// can be replaced with more useful operations.
+float ConsumerWork1(float f) {
+  float output = f;
+  for (int j = 0; j < kComplexity; j++) {
+    output = 20 * f + j - output;
+  }
+  return output;
+}
+
+float ConsumerWork2(float f) {
+  auto output = f;
+  for (int j = 0; j < kComplexity; j++) {
+    output = output + f * j;
+  }
+  return output;
+}
+
+/////////////////////////////////////////////////////////////
+// Pre-optimization kernel versions
+/////////////////////////////////////////////////////////////
+// The Producer kernel reads data from a SYCL buffer and writes it to
+// a pipe. This transfers the input data from the host to the Consumer kernel
+// that is running concurrently.
+// The Consumer kernel reads data from the pipe, performs the two ConsumerWork
+// operations on the data, and writes the results to the output buffer.
+
+void ProducerBefore(queue &q, buffer<float, 1>& buffer_a) {
+  auto e = q.submit([&](handler &h) {
+    // Get kernel access to the buffers
+    accessor accessor_a{buffer_a, h, read_only};
+
+    h.single_task<ProducerBeforeKernel>([=]() {
+      for (int i = 0; i < kSize; i++) {
+        ProducerToConsumerBeforePipe::write(accessor_a[i]);
+      }
+    });
+  });
+}
+
+void ConsumerBefore(queue &q, buffer<float, 1>& buffer_a) {
+  auto e = q.submit([&](handler &h) {
+    // Get kernel access to the buffers
+    accessor accessor_a{buffer_a, h, write_only, noinit};
+
+    h.single_task<ConsumerBeforeKernel>([=]() {
+      for (int i = 0; i < kSize; i++) {
+        auto input = ProducerToConsumerBeforePipe::read();
+        auto output = ConsumerWork1(input);
+        output = ConsumerWork2(output);
+        accessor_a[i] = output;
+      }
+    });
+  });
+}
+
+/////////////////////////////////////////////////////////////
+// Post-optimization kernel versions
+/////////////////////////////////////////////////////////////
+// The Producer kernel reads data from a SYCL buffer and does
+// ConsumerWork1 on it before giving it to the concurrently
+// running Consumer kernel.
+// The Consumer kernel reads data from the pipe, performs the rest
+// of the work (ConsumerWork2), and writes the results 
+// to the output buffer.
+
+void ProducerAfter(queue &q, buffer<float, 1>& buffer_a) {
+  auto e = q.submit([&](handler &h) {
+    // Get kernel access to the buffers
+    accessor accessor_a{buffer_a, h, read_only};
```

Comment:
Why do you need to use the "accessor_" prefix?   It seems just "a" would be simpler.  The strategy is choose your variable names to make the kernel easiest to read.